### ISSUE TITLE
Bus error 

### ISSUE THREAD

**Issue by wardpeet** [DoctorEvidence/lmdb-js#142]
Starting from version 2.2.0 we've been seeing bus errors coming up on Linux. Windows is fine and mac I haven't tested yet.

When we run it with `gdb node --ex "run ./node-modules/gatsby/cli.js build" we see that it's happening in lmdb
![image](https://user-images.githubusercontent.com/1120926/155352406-ec78a6fd-5214-4f05-b249-54ed0deb2e95.png)

Unsure if I can provide with more information

**Comment by wardpeet** [comment]
It seems to fail on v2.1.7 too so I'll see if I can find a way to properly reproduce it. I'm guessing it's happening on reads.

I'll come back to you in a bit

**Comment by wardpeet** [comment]
I think I nailed it down to `ifNoExists` method and it happens when we rapidly call it in succession. We're using it to create a mutex-like behavior. We're not using worker threads but we are using multiple processes who read/write to the same database.

Just to give you an idea we're calling it with 100 calls at a time with sometimes the same value. When I add a logging I can see that the last calls are all using the `ifNoExists`:

```
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:40f47410f0928cd73cfe499e28d4c225
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:75c56c20f33cee7ff5e15ca62f927e90
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:40f47410f0928cd73cfe499e28d4c225
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:1c3e8f837f26f64f072647fcd64a502b
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:75c56c20f33cee7ff5e15ca62f927e90
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:40f47410f0928cd73cfe499e28d4c225
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:501af5ee4ba0b55fe49722e5d1a94b02
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:40f47410f0928cd73cfe499e28d4c225
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:501af5ee4ba0b55fe49722e5d1a94b02
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:928a98a05a45773283c4d5c5938c5085
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:e1319c1810d70ceb3146edb6ce3f4ef9
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:1c3e8f837f26f64f072647fcd64a502b
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:75c56c20f33cee7ff5e15ca62f927e90
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:e1319c1810d70ceb3146edb6ce3f4ef9
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:501af5ee4ba0b55fe49722e5d1a94b02
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:75c56c20f33cee7ff5e15ca62f927e90
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:928a98a05a45773283c4d5c5938c5085
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:501af5ee4ba0b55fe49722e5d1a94b02
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:501af5ee4ba0b55fe49722e5d1a94b02
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:501af5ee4ba0b55fe49722e5d1a94b02
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:6a60878e94c7b42425ddb65a89e174c4
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:81653e2b1a1ec48f161670557c9fe00c
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:f6278196658535a423b3f280867decfe
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:81653e2b1a1ec48f161670557c9fe00c
[New Thread 0x7fff827fc700 (LWP 21371)]
[New Thread 0x7fff81ffb700 (LWP 21372)]
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:f6278196658535a423b3f280867decfe
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:81653e2b1a1ec48f161670557c9fe00c
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:81653e2b1a1ec48f161670557c9fe00c
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:1462eae5eb753c2e12fb212e52095a33
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:81653e2b1a1ec48f161670557c9fe00c
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:81653e2b1a1ec48f161670557c9fe00c
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:1462eae5eb753c2e12fb212e52095a33
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:075ce78d7c0d03e414b5f5b934eb4638
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:4b839471adc08f729e74db24ed3abd3e
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:075ce78d7c0d03e414b5f5b934eb4638
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:4b839471adc08f729e74db24ed3abd3e
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:98dc09a1a3b1888e9141d6a41880a5e5
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:811c6f6c08a64d1d64ef973ae78dc7ff
lock 62d27031-c4e8-41fa-a299-467b15cc52ec-gatsby-plugin-utils:placeholder:98dc09a1a3b1888e9141d6a41880a5e5
[Thread 0x7fffb20bd700 (LWP 21365) exited]
[Thread 0x7fffb28be700 (LWP 21349) exited]
[Thread 0x7fff817fa700 (LWP 21360) exited]

Thread 8 "node" received signal SIGBUS, Bus error.
```

<details>
<summary>Db initialization code</summary>

```
import path from "path"
import { getLmdb } from "./get-lmdb"
import type { RootDatabase, Database } from "lmdb"
import type { Headers } from "got"

export enum LockStatus {
  Locked = 0,
  Unlocked = 1,
}

interface ICoreUtilsDatabase {
  remoteFileInfo: Database<
    {
      extension: string
      headers: Headers
      path: string
      directory: string
      cacheKey?: string
      buildId: string
    },
    string
  >
  mutex: Database<LockStatus, string>
}

let databases: ICoreUtilsDatabase | undefined
let rootDb: RootDatabase

export function getDatabaseDir(): string {
  const rootDir = global.__GATSBY?.root ?? process.cwd()
  return path.join(rootDir, `.cache`, `data`, `gatsby-core-utils`)
}

export function getStorage(fullDbPath: string): ICoreUtilsDatabase {
  if (!databases) {
    if (!fullDbPath) {
      throw new Error(`LMDB path is not set!`)
    }

    // __GATSBY_OPEN_LMDBS tracks if we already opened given db in this process
    // In `gatsby serve` case we might try to open it twice - once for engines
    // and second to get access to `SitePage` nodes (to power trailing slashes
    // redirect middleware). This ensure there is single instance within a process.
    // Using more instances seems to cause weird random errors.
    if (!globalThis.__GATSBY_OPEN_LMDBS) {
      globalThis.__GATSBY_OPEN_LMDBS = new Map()
    }

    databases = globalThis.__GATSBY_OPEN_LMDBS.get(fullDbPath)

    if (databases) {
      return databases
    }

    const open = getLmdb().open

    rootDb = open({
      name: `root`,
      path: fullDbPath,
      compression: true,
      sharedStructuresKey: Symbol.for(`structures`),
    })

    databases = {
      remoteFileInfo: rootDb.openDB({
        name: `remote-file`,
      }),
      mutex: rootDb.openDB({
        name: `mutex`,
      }),
    }

    globalThis.__GATSBY_OPEN_LMDBS.set(fullDbPath, databases)
  }

  return databases as ICoreUtilsDatabase
}

export async function closeDatabase(): Promise<void> {
  if (rootDb) {
    await rootDb.close()
  }
}
```
</details>

<details>
<summary>mutex code</summary>

```
import { getStorage, LockStatus, getDatabaseDir } from "./utils/get-storage"

interface IMutex {
  acquire(): Promise<void>
  release(): Promise<void>
}

// Random number to re-check if mutex got released
const DEFAULT_MUTEX_INTERVAL = 3000

async function waitUntilUnlocked(
  storage: ReturnType<typeof getStorage>,
  key: string,
  timeout: number
): Promise<void> {
  const isUnlocked = await storage.mutex.ifNoExists(key, () => {
    storage.mutex.put(key, LockStatus.Locked)
  })

  if (isUnlocked) {
    return
  }

  await new Promise<void>(resolve => {
    setTimeout(() => {
      resolve(waitUntilUnlocked(storage, key, timeout))
    }, timeout)
  })
}

/**
 * Creates a mutex, make sure to call `release` when you're done with it.
 *
 * @param {string} key A unique key
 */
export function createMutex(
  key: string,
  timeout = DEFAULT_MUTEX_INTERVAL
): IMutex {
  const storage = getStorage(getDatabaseDir())
  const BUILD_ID = global.__GATSBY?.buildId ?? ``
  const prefixedKey = `${BUILD_ID}-${key}`

  return {
    acquire: (): Promise<void> =>
      waitUntilUnlocked(storage, prefixedKey, timeout),
    release: async (): Promise<void> => {
      await storage.mutex.remove(prefixedKey)
    },
  }
}

export async function releaseAllMutexes(): Promise<void> {
  const storage = getStorage(getDatabaseDir())

  await storage.mutex.clearAsync()
}

```
</details>

**Comment by kriszyp** [comment]
Thank you for the detailed code samples. Unfortunately, I still haven't been able to reproduce this; here is my simple, naive attempt: https://github.com/DoctorEvidence/lmdb-js/commit/f24645cfc6a51d52e6dc734cfdf8b56fdbe9919c#diff-af95cb6f5f65fe70216ddce70325e785438d5aaff956a0666ebf46ef6402f526R261
Perhaps multiple processes are important in this too though.
A few other questions:
Is this on a typical x64 architecture (that would generally rule out actual memory alignment issues)? I have read that this signal can also be caused by disk I/O issues with memory mapped files. Is it possible this could be caused by running out of space, or any other type of disk I/O problem? Is this a standard disk drive or a network file system of some sort?

**Comment by wardpeet** [comment]
This reproduction fails for me 99% of the time
```
const lmdb = require('lmdb');
const fs = require('fs');

try {
  fs.rmSync('./database/data.mdb');
  fs.rmSync('./database/lock.mdb');
  fs.rmdirSync('./database');
} catch(err) {
  // ignore
}

const database = lmdb.open({
  name: `root`,
  path: './database',
  compression: true,
  sharedStructuresKey: Symbol.for(`structures`),
});


(async () => {
  const key = '1';
  for (let i = 0; i < 100; i++) {
    await database.ifNoExists(key, () => {
      database.put(key, 1)
    })

    await new Promise(resolve => {
      setTimeout(() => resolve(), 100 * Math.random())
    })

    await database.remove(key)
  }
})();
```

**Comment by kriszyp** [comment]
Great, thank you for the perfect test case! Should be fixed in v2.2.3.
